main/src/ca/uwaterloo/flix/language/phase/njvm/Mnemonics.scala

```diff
@@ -89,4 +150,118 @@ object Mnemonics {
 
   }
 
+  //Capabilities
+  /**
+    * Capability which allows to load/store a local variable
+    */
+  sealed trait Local[T]{
+
+    val localType : JvmType
+    val location : Int
+
+    def LOAD[R <: Stack]() : F[R] => F[R**T] =  t => t.load(localType, location)
+    def STORE[R <: Stack]() : F[R**T] => F[R] =  t => t.load(localType, location)
+  }
+
+  class LocalC[T](val localType: JvmType, val location: Int) extends Local[T]
+
+  /**
+    * Capability which allows to get/put a field
+    */
+  sealed trait Field[T]{
+    val fieldName : String
+    val fieldType : JvmType
+
+    def GET_FIELD[R <: Stack]() : F[R ** JvmType.Reference] => F[R ** T] =  t => t.getField(fieldName, fieldType)
+    def PUT_FIELD[R <: Stack]() : F[R ** JvmType.Reference ** T] => F[R] =  t => t.putField(fieldName, fieldType)
+  }
+
+  class FieldC[T](val fieldName: String, val fieldType: JvmType) extends Field[T]
+
+  sealed trait FunSig[R] {
+    val modifiers : List[JvmModifier]
+    val methodName : String
+    val returnType : JvmType
+
+    def getArgsType : List[JvmType]
+  }
+
+  class FunSig0[R](val modifiers : List[JvmModifier], val methodName : String, val returnType : JvmType,
+                   classType: JvmType) extends FunSig[R] {
+   val arg0 : Local[JvmType.Reference] = new LocalC[JvmType.Reference](classType, 0)
+
+    override def getArgsType: List[JvmType] = List()
+  }
+
+  class FunSig1[T1, R](val modifiers : List[JvmModifier], val methodName : String, val returnType : JvmType,
+                       classType: JvmType, arg1Type: JvmType) extends FunSig[R] {
+    val arg0 : Local[JvmType.Reference] = new LocalC[JvmType.Reference](classType, 0)
+    val arg1 : Local[T1] = new LocalC[T1](arg1Type, 1)
+
+    override def getArgsType: List[JvmType] = List(arg1Type)
+  }
+
+
+  trait MethodHandler0Void {
+    val invokeCode : JvmModifier
+    val className : String
+    val methodName: String
+    val args : List[JvmType]
+    val returnType : JvmType
+
+    def invoke[R <: Stack ](): F[R ** JvmType.Reference] => F[R] =
+      t => t.invoke(invokeCode, className, methodName, args, returnType)
+  }
+
+  class MethodHandler0VoidC(val invokeCode : JvmModifier, val  className : String, val methodName: String,
+                            val args : List[JvmType], val returnType : JvmType) extends MethodHandler0Void
+
+
+  def mkMethodHandler0Void[R <: Stack](invokeCode : JvmModifier, className : String, methodName: String,
+                                       args : List[JvmType], returnType : JvmType) : MethodHandler0Void=
+  { new MethodHandler0VoidC(invokeCode,className, methodName, args, returnType) }
+
+
+  class ClassGenerator(classType: JvmType.Reference,
+                       modifiers : List[JvmModifier], superClass : String, implementedInterfaces : Array[String])
```

Comment:
Maybe the modifiers should be a set?
Also, does it make sense to split the modifiers of classes and methods? Yes/no?

Reply:
superClass is a string... Not good. How about a JvmType.Reference at least?

Reply:
Same comment for implementedInterfaces. Also, why is this an array?